covalent_indexer.go

```diff
@@ -1,34 +1,90 @@
 package covalent
 
 import (
+	"encoding/hex"
+	"net/http"
+	"sync"
+
+	"github.com/ElrondNetwork/covalent-indexer-go/process/utility"
+	"github.com/ElrondNetwork/covalent-indexer-go/process/ws"
+	"github.com/ElrondNetwork/covalent-indexer-go/schema"
 	"github.com/ElrondNetwork/elrond-go-core/data"
 	"github.com/ElrondNetwork/elrond-go-core/data/indexer"
+	logger "github.com/ElrondNetwork/elrond-go-logger"
 )
 
+var log = logger.GetOrCreate("covalent")
+
 type covalentIndexer struct {
 	processor DataHandler
+	server    *http.Server
+	wss       *ws.WSSender
+	sync.RWMutex
 }
 
 // NewCovalentDataIndexer creates a new instance of covalent data indexer, which implements Driver interface and
 // converts protocol input data to covalent required data
-func NewCovalentDataIndexer(processor DataHandler) (Driver, error) {
-	return &covalentIndexer{
+func NewCovalentDataIndexer(processor DataHandler, server *http.Server) (*covalentIndexer, error) {
+	if processor == nil {
+		return nil, ErrNilDataHandler
+	}
+	if server == nil {
+		return nil, ErrNilHTTPServer
+	}
+
+	ci := &covalentIndexer{
 		processor: processor,
-	}, nil
+		server:    server,
+	}
+
+	go ci.start()
+
+	return ci, nil
+}
+
+func (c *covalentIndexer) SetWSSender(wss *ws.WSSender) {
+	c.Lock()
+	defer c.Unlock()
+	if c.wss != nil {
+		_ = c.wss.Conn.Close()
+	}
+	c.wss = wss
+}
+
+func (c *covalentIndexer) start() {
+	err := c.server.ListenAndServe()
+	if err != nil {
+		log.Error("could not initialize webserver", "error", err)
+	}
+}
+
+func (c *covalentIndexer) sendBlockResultToCovalent(result *schema.BlockResult) {
+	binaryData, err := utility.Encode(result)
+	if err != nil {
+		log.Error("could not encode block result to binary data", "error", err)
+		return
+	}
+
+	c.RLock()
+	wss := c.wss
```

Comment:
why you need this  ? you just pass the pointer address

Reply:
When working with sockets (sender/receiver), I want to use the existing connections (by locking them). In case a new connection appears (e.g.: receiver got disconnected and reconnects) the mechanism will try to send data via the old(disconnected) connection -> will realize the connection is compromised -> will listen again to new connection

Reply:
I want to use the actual pointer, not a copy